File: src/main/java/dev/zenhao/melon/utils/BackgroundEffect/FakeMeteor/Meteor.java
package dev.zenhao.melon.utils.BackgroundEffect.FakeMeteor;

import dev.zenhao.melon.utils.Rainbow;
import dev.zenhao.melon.utils.TimerUtils;
import dev.zenhao.melon.utils.render.FadeUtils;
import org.lwjgl.opengl.Display;
import org.lwjgl.util.vector.Vector2f;

import java.awt.*;
import java.util.Random;

public class Meteor {
    public static Random random = new Random();
    private final TimerUtils timerUtils = new TimerUtils();
    private final FadeUtils fadeUtils = new FadeUtils(170L);
    public Vector2f pos, pos2;
    public float LineWidth, LineLength, alpha;
    public long speedMS;
    public Color randomColor;

    public Meteor(long speedMS, float x, float y, float LineWidth, float LineLength) {
        this.speedMS = speedMS;
        this.pos = new Vector2f(x, y);
        this.pos2 = new Vector2f(x, y);
        this.LineWidth = LineWidth;
        this.LineLength = LineLength;
        this.randomColor = new Color(Color.HSBtoRGB(random.nextInt(360), 0.4f, 1.0f));
        fadeUtils.setLength((speedMS / 5) * 2);
    }

    public static Meteor generateMeteor() {
        long speedMS = 3000 + random.nextInt(1200);
        float x = random.nextInt(Display.getWidth());
        float y = random.nextInt(Display.getHeight());
        float lineLength = 50 + random.nextInt(300);
        float lineWidth = (float) (Math.random() * 2.0f) + 1.0f;
        return new Meteor(speedMS, x, y, lineWidth, lineLength);
    }

    public float getAlpha() {
        return this.alpha;
    }

    public float getLineWidth() {
        return this.LineWidth;
    }

    public void setLineWidth(float f) {
        this.LineWidth = f;
    }

    public float getX() {
        return this.pos.x;
    }

    public float getY() {
        return this.pos.y;
    }

    public float getX2() {
        return this.pos2.x;
    }

    public float getY2() {
        return this.pos2.y;
    }

    public void tick() {
        if (timerUtils.passed(speedMS)) {
            timerUtils.reset();
            pos.x = (random.nextInt(Display.getWidth()));
            pos.y = (random.nextInt(Display.getHeight()));
            LineLength = 70 + random.nextInt(300);
            randomColor = new Color(Rainbow.getRainbow(random.nextInt(360), 0.4f, 1.0f));
            fadeUtils.reset();
            alpha = 0;
        }

        double speedMoves = speedMS / 5f;
        if (timerUtils.passed(speedMoves * 3)) {
            pos.x = (float) (pos2.x + (LineLength * fadeUtils.getFade(FadeUtils.FadeMode.FADE_EASE_OUT_QUAD)));
            pos.y = (float) (pos2.y - (LineLength * fadeUtils.getFade(FadeUtils.FadeMode.FADE_EASE_OUT_QUAD)));
        } else if (timerUtils.passed(speedMoves * 2)) {
            fadeUtils.reset();
        } else {
            pos2.x = (float) (pos.x - (LineLength * fadeUtils.getFade(FadeUtils.FadeMode.FADE_EASE_IN_QUAD)));
            pos2.y = (float) (pos.y + (LineLength * fadeUtils.getFade(FadeUtils.FadeMode.FADE_EASE_IN_QUAD)));
        }

        /*

        if (timerUtils.passed(speedMoves * 3)) {
            pos.x = (float) (pos2.x + (LineLength * fadeUtils.getFade(FadeUtils.FadeMode.FADE_OUT)));
            pos.y = (float) (pos2.y - (LineLength * fadeUtils.getFade(FadeUtils.FadeMode.FADE_OUT)));
        } else if (timerUtils.passed(speedMoves * 2)) {
            fadeUtils.reset();
        } else {
            pos2.x = (float) (pos.x - (LineLength * fadeUtils.getFade(FadeUtils.FadeMode.FADE_IN)));
            pos2.y = (float) (pos.y + (LineLength * fadeUtils.getFade(FadeUtils.FadeMode.FADE_IN)));
        }
         */

        if (this.alpha < 255.0f) {
            this.alpha += 15f;
        }
    }
}

File: src/main/kotlin/melon/system/render/background/particle/Meteor.kt
package melon.system.render.background.particle

import melon.system.render.background.AParticle
import melon.system.render.background.particle.Meteor.MeteorState.*
import dev.zenhao.melon.utils.TickTimer
import dev.zenhao.melon.utils.TimeUnit
import dev.zenhao.melon.utils.animations.AnimationFlag
import dev.zenhao.melon.utils.animations.Easing
import dev.zenhao.melon.utils.render.FadeUtils
import dev.zenhao.melon.utils.render.RenderUtils
import dev.zenhao.melon.utils.vector.Vec2f
import org.lwjgl.opengl.Display
import java.awt.Color
import kotlin.random.Random

class Meteor(pos: Vec2f, size: Float, velocity: Vec2f, speed: Long, color: Color) : AParticle(pos, size, velocity, speed, color) {

    var alpha = color.alpha.toFloat()
        set(value) {
            field = value.coerceAtMost(255F)
        }

    private val refreshTimer = TickTimer(TimeUnit.MILLISECONDS)

    private var state = IN

    private val easing get() = if (state == OUT) Easing.OUT_QUAD else Easing.IN_QUAD

    override fun render() {
        val speedMoves = speed / 5L
        state = if (refreshTimer.tick(speedMoves * 3L)) OUT else if (refreshTimer.tick(speedMoves * 2L)) STAY else IN
        val offset = easing.inc(Easing.toDelta(refreshTimer.time + if (state == OUT) speedMoves * 3L else 0L, speedMoves * 2F), velocity.x)

        val from: Vec2f
        val to: Vec2f
        when (state) {
            IN, STAY -> {
                from = pos
                to = pos + Vec2f(-offset, offset)
            }
            OUT -> {
                from = pos + Vec2f(-velocity.x, velocity.x)
                to = pos + Vec2f(-offset, offset)
            }
        }

        RenderUtils.drawLine(
            from.x.toDouble(),
            from.y.toDouble(),
            to.x.toDouble(),
            to.y.toDouble(),
            size,
            Color(color.red, color.green, color.blue, alpha.toInt())
        )
    }

    override fun refresh() {
        pos = Vec2f(Random.nextInt(Display.getWidth()).toFloat(), Random.nextInt(Display.getHeight()).toFloat())
        velocity = Vec2f(50F + Random.nextInt(300), 0F)
        color = Color(Color.HSBtoRGB(Random.nextInt(360).toFloat(), 0.4f, 1.0f))
        alpha = if (color.alpha == 255) 15F else color.alpha.toFloat()
    }

    override fun update() {
        if (refreshTimer.tickAndReset(speed)) refresh()
        if (alpha < 255F) this.alpha += 0.05f * 10
    }

    override fun generate(): AParticle = Companion.generate()

    companion object {

        @JvmField
        val GET: Meteor = generate()


        @JvmStatic
        fun generate(): Meteor {
            val pos = Vec2f(Random.nextInt(Display.getWidth()).toFloat(), Random.nextInt(Display.getHeight()).toFloat())
            val size = Math.random().toFloat() * 3.0f + 1.0f
            val velocity = Vec2f(50F + Random.nextInt(300), 0F)
            val color = Color(Color.HSBtoRGB(Random.nextInt(360).toFloat(), 0.4f, 1.0f))
            val speed = 3000L + Random.nextInt(1200)
            return Meteor(pos, size, velocity, speed, color)
        }
    }


    private enum class MeteorState {
        IN,
        STAY,
        OUT
    }
}